src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
                 }
 #endif // FEATURE_EH_FUNCLETS && defined(_TARGET_ARM_)
 
-                /* At this point we have a loop - record it in the loop table
+                // Compact the loop (sweep through it and move out any blocks that aren't part of the
+                // flow cycle), and count the exits.
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
+
+                    if (!loopBlocks.isMember(block->bbNum))
+                    {
+                        // This code is lexically between TOP and BOTTOM, but it does not
+                        // participate in the flow cycle.  Check for a run of consecutive
+                        // such blocks.
+                        BasicBlock* lastExitBlock = block;
+                        BasicBlock* nextLoopBlock = block->bbNext;
+                        while (!loopBlocks.isMember(nextLoopBlock->bbNum))
+                        {
+                            lastExitBlock = nextLoopBlock;
+                            nextLoopBlock = nextLoopBlock->bbNext;
+                            // This loop must terminate because we know BOTTOM is in loopBlocks.
+                        }
+
+                        if (moveAfter == nullptr)
+                        {
+                            // Find an insertion point for blocks we're going to move.  Move them down
+                            // out of the loop, and if possible find a spot that won't break up fall-through.
+                            moveAfter = bottom;
+                            while (moveAfter->bbFallsThrough())
+                            {
+                                // Avoid splitting up this fall-through if we can.
+                                BasicBlock* next = moveAfter->bbNext;
+
+                                if (!BasicBlock::sameEHRegion(moveAfter, next))
+                                {
+                                    // Don't cross an EH region boundary
+                                    break;
+                                }
+
+                                if ((next->bbJumpKind == BBJ_ALWAYS) || (next->bbJumpKind == BBJ_COND))
+                                {
+                                    unsigned int destNum = next->bbJumpDest->bbNum;
+                                    if ((destNum >= top->bbNum) && (destNum <= bottom->bbNum) &&
+                                        !loopBlocks.isMember(destNum))
+                                    {
+                                        // Reversing this branch out of block `next` could confuse this algorithm, so
```

Comment:
> Maybe you just need to be a bit more explicit on what exactly about reversal leads to confusion.

The algorithm expects that if it can find an edge from a block with higher number to a block with lower number, that it has found a lexically backward edge and therefore that walking `bbNext` pointers from the successor will find the predecessor.  It also assumes, when walking the loop blocks, that any block it visits with number between TOP and BOTTOM is lexically between the two. 
 Performing any motion that reverses the lexical direction of a flow edge risks breaking those invariants.
(and the motion we do perform can mean that blocks lexically between TOP and BOTTOM in a subsequent loop have numbers less than TOP's number, but I believe that can only cause conservativism [we'll treat flow into or out of that region as exits and side-entries when we maybe didn't need to], not correctness issues)

> avoid creating new lexically backwards branches

I think that's a correctness constraint, and that's what this code is trying to do.

> Insert after BOTTOM would guarantee this, though ...

I started with always inserting just after bottom, but the diffs showed lots of cases where that led to ugly layout; in particular, given a small method with a search loop that has an early return and the normal loop exit falls through to the other return, it unnecessarily puts an unconditional jump on the normal exit path.

> you could get the same guarantee if you moved the run of blocks before any post-loop block they reach, and if there is just one such block and maybe try and move the run just before it.

That's really what this code is trying to do.  And since I can't guarantee correctness if I move past any such block, the search stops at the first such block and we move the run just before it.


> I probably would have tried to leverage `fgFindInsertPoint` to do the actual searching, and looking at it now, it seems like there might be logic there you need here, eg a `BBJ_CALLFINALLY` does not fall through but you can't safely move new blocks after it.

Thanks, I didn't know about `fgFindInsertPoint` or that constraint; I'll have to take a look and update...